mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,694 @@
+        """ Extracts entities from a sentence. Detected entities should are
+        represented as dictionaries with the following keys: "body", "start"
+        (start index), "end" (end index), "value", "dim" (entity type).
+
+        Args:
+            sentence (str): Sentence to detect entities.
+            entity_types (list): List of entity types to parse. If None, all
+                possible entity types will be parsed.
+
+        Returns:
+            entities (list): List of entity dictionaries.
+        """
+        doc = self.nlp(sentence)
+        spacy_entities = [
+            {
+                "body": ent.text,
+                "start": ent.start_char,
+                "end": ent.end_char,
+                "value": {"value": ent.text},
+                "dim": ent.label_.lower(),
+            }
+            for ent in doc.ents
+        ]
+
+        entities = []
+        for entity in spacy_entities:
+            if entity["dim"] in ["time", "date"]:
+                entity = self._resolve_time_date(entity, entity_types)
+            elif entity["dim"] == "cardinal":
+                entity = self._resolve_cardinal(entity)
+            elif entity["dim"] == "money":
+                entity = self._resolve_money(entity, sentence)
+            elif entity["dim"] == "ordinal":
+                entity = self._resolve_ordinal(entity)
+            elif entity["dim"] == "quantity":
+                entity = self._resolve_quantity(entity)
+            elif entity["dim"] == "percent":
+                entity = self._resolve_percent(entity)
+            elif entity["dim"] == "person":
+                entity = self._resolve_person(entity)
+            else:
+                entity["dim"] = "sys_" + entity["dim"].replace("_", "-")
+
+            if entity:
+                entities.append(entity)
+
+        if entity_types:
+            entities = [e for e in entities if e["dim"] in entity_types]
+
+        return entities
+
+    def _resolve_time_date(self, entity, entity_types=None):
+        """ Resolves a time related entity. First looks for an exact match, then
+        for the largest substring match. Order of priority is sys_duration, sys_interval,
+        and sys_time.
+
+        Args:
+            entity (dict): A dictionary representing an entity.
+            entity_types (list): List of entity types to parse. If None, all possible
+                entity types will be parsed.
+
+        Returns:
+            entity (dict): A resolved entity dict or None if the entity isn't resolved.
+        """
+        candidates = self.duckling.get_candidates_for_text(entity["body"])
+
+        if len(candidates) == 0:
+            return
+
+        time_entities = ["sys_duration", "sys_interval", "sys_time"]
+        if entity_types:
+            time_entities = [e for e in time_entities if e in entity_types]
+
+        if SpacyAnnotator._resolve_time_exact_match(entity, candidates, time_entities):
```

Comment:
Add an explanation on why the conditionals are done in this order.

Reply:
Added